pkg/autodiscovery/configmgr.go

```diff
@@ -0,0 +1,292 @@
+		tpls, err := cm.store.templateCache.get(adID)
+		if err != nil {
+			log.Debugf("Unable to fetch templates from the cache: %v", err)
+		}
+		templates = append(templates, tpls...)
+	}
+
+	resolvedSet := map[string]integration.Config{}
+	for _, template := range templates {
+		// resolve the template
+		resolvedConfig, err := cm.resolveTemplateForService(template, svc)
+		if err != nil {
+			continue
+		}
+
+		resolvedSet[resolvedConfig.Digest()] = resolvedConfig
+	}
+
+	// build the config changes to return
+	changes := configChanges{}
+	for _, v := range resolvedSet {
+		changes.scheduleConfig(v)
+	}
+
+	return changes
+}
+
+// processDelService implements configManager#processDelService.
+func (cm *simpleConfigManager) processDelService(svc listeners.Service) configChanges {
+	cm.m.Lock()
+	defer cm.m.Unlock()
+	changes := configChanges{}
+
+	cm.store.removeServiceForEntity(svc.GetServiceID())
+	removedConfigs := cm.store.removeConfigsForService(svc.GetServiceID())
+	for _, c := range removedConfigs {
+		if cm.store.removeLoadedConfig(c) {
+			changes.unscheduleConfig(c)
+		}
+	}
+
+	return changes
+}
+
+// processNewConfig implements configManager#processNewConfig.
+func (cm *simpleConfigManager) processNewConfig(config integration.Config) configChanges {
+	cm.m.Lock()
+	defer cm.m.Unlock()
+	changes := configChanges{}
+
+	if config.IsTemplate() {
+		// store the template in the cache in any case
+		if err := cm.store.templateCache.set(config); err != nil {
+			log.Errorf("Unable to store Check configuration in the cache: %s", err)
+		}
+
+		// try to resolve the template
+		resolvedConfigs := cm.resolveTemplate(config)
+		if resolvedConfigs.isEmpty() {
+			e := fmt.Sprintf("Can't resolve the template for %s at this moment.", config.Name)
+			errorStats.setResolveWarning(config.Name, e)
+			log.Debug(e)
+			return changes // empty result
+		}
+
+		return resolvedConfigs
+	}
+
+	// decrypt and store non-template config in AC as well
+	config, err := decryptConfig(config)
+	if err != nil {
+		log.Errorf("Dropping conf for '%s': %s", config.Name, err.Error())
+		return changes // empty result
+	}
+	changes.scheduleConfig(config)
+	cm.store.setLoadedConfig(config)
+
+	return changes
+}
+
+// processDelConfigs implements configManager#processDelConfigs.
+func (cm *simpleConfigManager) processDelConfigs(configs []integration.Config) configChanges {
+	cm.m.Lock()
+	defer cm.m.Unlock()
+	changes := configChanges{}
+
+	for _, c := range configs {
+		if c.IsTemplate() {
+			// Remove the resolved configurations
+			tplDigest := c.Digest()
+			removedConfigs := cm.store.removeConfigsForTemplate(tplDigest)
+			for _, rc := range removedConfigs {
+				if cm.store.removeLoadedConfig(rc) {
+					changes.unscheduleConfig(rc)
+				}
+			}
+
+			// Remove template from the cache
+			err := cm.store.templateCache.del(c)
+			if err != nil {
+				log.Debugf("Could not delete template: %v", err)
+			}
+		} else {
+			cm.store.removeLoadedConfig(c)
```

Comment:
I added that to solve an issue of doubly-unscheduling a resolved config (once when its template is removed, and again when its service is removed).  In this particular case, it's not important to check because this s a non-template config and not subject to resolution.

The fuzzing would catch any issues with this (that's how I found the double-unscheduling bug)